feat(examples,metrics,kube-state-metrics): add configmap and promethe…

sebastiangaiser opened the pull request:
…us-rules

In order to implement https://github.com/strimzi/proposals/blob/main/087-monitoring-of-custom-resources.md an initial version of the configmap and prometheus-rules are added for 'KafkaUser' and 'KafkaTopic' resources.

Part of https://github.com/strimzi/strimzi-kafka-operator/issues/10276

### Type of change
- Enhancement / new feature

### Description

See commit message

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [x] Supply screenshots for visual changes, such as Grafana dashboards

sebastiangaiser commented:
I'm unsure wehter `KafkaTopic` could also have a `DeprecatedFields` status field.

sebastiangaiser commented on packaging/examples/metrics/kube-state-metrics/configmap.yaml:
```diff
@@ -0,0 +1,50 @@
+                exported_namespace: [ metadata, namespace ]
+                partitions: [ spec, partitions ]
+                replicas: [ spec, replicas ]
+                ready: [ status, conditions, "[type=Ready]", status ]
```
```suggestion
                ready: [ status, conditions, "[type=Ready]", status ]
                deprecated: [ status, conditions, "[reason=DeprecatedFields]", type ]
```

sebastiangaiser commented on packaging/examples/metrics/kube-state-metrics/prometheus-rules.yaml:
```diff
@@ -0,0 +1,30 @@
+            severity: warning
+          annotations:
+            message: "Strimzi KafkaUser {{`{{ $labels.username }}`}} is not ready"
+        - alert: KafkaTopicNotReady
```
```suggestion
        - alert: KafkaUserDeprecated
          expr: strimzi_kafka_topic_resource_info{deprecated="Warning"}
          for: 15m
          labels:
            severity: warning
          annotations:
            message: "Strimzi KafkaTopic {{`{{ $labels.topicName }}`}} has a deprecated configuration"
        - alert: KafkaTopicNotReady
```

sebastiangaiser commented on packaging/examples/metrics/kube-state-metrics/prometheus-rules.yaml:
```diff
@@ -0,0 +1,37 @@
+            severity: warning
+          annotations:
+            message: "Strimzi KafkaUser {{`{{ $labels.username }}`}} is not ready"
+        - alert: KafkaUserDeprecated
```
```suggestion
        - alert: KafkaTopicDeprecated
```

sebastiangaiser commented on packaging/examples/metrics/kube-state-metrics/prometheus-rules.yaml:
```diff
@@ -0,0 +1,37 @@
+          labels:
+            severity: warning
+          annotations:
+            message: "Strimzi Kafka Topic {{`{{ $labels.topicName }}`}} is not ready"
```
```suggestion
            message: "Strimzi KafkaTopic {{`{{ $labels.topicName }}`}} is not ready"
```

scholzj reviewed (commented):
I think this looks good. But it would be great to probably:
* Have some README.md file here explaining what this is and how to use it?
* Maybe we can add some simple chapter to the end of this section in the docs (https://strimzi.io/docs/operators/latest/full/deploying.html#assembly-metrics-str) that would at least mention this as well?

ppatierno commented:
I had a first pass and agree with Jakub on what's missing. I will have another pass when the upgrades are committed.

sebastiangaiser commented:
I've tried to address your points: 
- added a README in the example directory
- added a section to the documentation. Tbh I've never worked with AsciiDoc before, so please guide me a bit here ;)

In addition I added a KSM deployment (similar as it already exists for Grafana) for clarifying how a Kubernetes deployment could look like.

scholzj reviewed (approved):
LGTM. Thanks.

scholzj commented:
@ppatierno @PaulRMellor Could you please have a look at this?

ppatierno commented on documentation/assemblies/metrics/assembly-metrics.adoc:
```diff
@@ -59,4 +59,6 @@ include::assembly-metrics-config-files.adoc[leveloffset=+1]
-include::../../modules/metrics/proc_metrics-grafana-dashboard.adoc[leveloffset=+1]
\ No newline at end of file
+include::../../modules/metrics/proc_metrics-grafana-dashboard.adoc[leveloffset=+1]
+//How to montitor custom resources managed by Strimzi
```
```suggestion
//How to monitor custom resources managed by Strimzi
```

ppatierno commented on documentation/modules/metrics/proc_metrics-custom-resource-monitoring.adoc:
```diff
@@ -0,0 +1,68 @@
+kubectl port-forward svc/strimzi-kube-state-metrics 8080:8080
+----
+
+. In a web browser, access the KSM metrics page using the URL `http://localhost:8080/metrics`.
```
the specific above steps related to the port-forward are referring to a deployment using minikube or something similar I guess. They should be not needed if the user is using a Kubernetes cluster somewhere else (i.e. AKS, ...). Not sure we want this part.

sebastiangaiser replied:
Part is removed.

ppatierno commented on documentation/modules/metrics/proc_metrics-custom-resource-monitoring.adoc:
```diff
@@ -0,0 +1,68 @@
+You can use your own KSM deployment or deploy KSM using the xref:assembly-metrics-config-files-{context}[example metrics configuration files] provided by Strimzi.
+The example files include a configuration file for a KSM deployment
+
+* `examples/metrics/kube-state-metrics/ksm.yaml`
```
AFAICS this YAML contains some cluster role binding with a specific namespace "myproject". If the user is deploying in a different one, it's not going to work, or? 
We have the same issue when installing the operator but we provide the `sed` command to replace the namespace with the right one. Should we write something similar here?

sebastiangaiser replied:
I've taken some inspiration from the Prometheus installation and added some lines. In addition I changed the namespace to have a similar `sed` command.

scholzj replied:
I think the sed command is good. But the namespace in the files should probably default to `myproject` as that is what we use in the other YAMLs by default.

sebastiangaiser replied:
Thanks. Updated 👍🏻

ppatierno reviewed (approved):
LGTM. Thanks!

ppatierno commented:
@PaulRMellor is there anything to address here from your pov?

PaulRMellor commented:
> @PaulRMellor is there anything to address here from your pov?

Going to review now

PaulRMellor commented on documentation/modules/metrics/proc_metrics-custom-resource-monitoring.adoc:
```diff
@@ -0,0 +1,53 @@
+kubectl apply -f configmap.yaml
+kubectl apply -f ksm.yaml
+
+After deploying KSM, you can check that all of these metrics get scraped by Prometheus via the provided `ServiceMonitor` in order to act on in Prometheus.
+
+Please check the provided `PrometheusRule` resources for alerting on these metrics:
+
+* `examples/metrics/kube-state-metrics/prometheus-rules.yaml`
```
```suggestion
. Verify that Prometheus is scraping KSM metrics.
+
Metrics are scraped from the `strimzi-kube-state-metrics` service using the `ServiceMonitor` configured for the KSM deployment.
+
For alerting on these metrics, check the provided `PrometheusRule` resources:
+
* `examples/metrics/kube-state-metrics/prometheus-rules.yaml`
```

PaulRMellor replied:
I thought this could be a second step

PaulRMellor reviewed (commented):
Thanks for the addition. Looks great. I left a few suggestions for the doc.

sebastiangaiser commented:
@PaulRMellor thank you for reviewing the changes. I simply applied your comments. Please let me know if something is missing.

PaulRMellor reviewed (approved):
Excellent. Thanks for the doc updates
